stdlib/public/Concurrency/AsyncStreamBuffer.swift

```diff
@@ -212,25 +212,31 @@ extension AsyncStream {
       state.onTermination = nil
       state.terminal = true
 
-      if let continuation = state.continuations.first {
-        if state.pending.count > 0 {
-          state.continuations.removeFirst()
-          let toSend = state.pending.removeFirst()
-          unlock()
-          handler?(.finished)
-          continuation.resume(returning: toSend)
-        } else if state.terminal {
-          state.continuations.removeFirst()
-          unlock()
-          handler?(.finished)
-          continuation.resume(returning: nil)
-        } else {
-          unlock()
-          handler?(.finished)
-        }
-      } else {
+      guard !state.continuations.isEmpty else {
         unlock()
         handler?(.finished)
+        return
+      }
+
+      // Get elements to deliver to outstanding continuations.
+      let upperBound = Swift.min(state.continuations.count, state.pending.count)
+      let elementsRange = 0..<upperBound
+      let elementsToSend = state.pending[elementsRange]
+      state.pending.removeSubrange(elementsRange)
+
+      // Hold on to the continuations to resume outside the lock.
+      let continuations = state.continuations
+      state.continuations = []
+
+      unlock()
+      handler?(.finished)
+
+      for (idx, continuation) in continuations.enumerated() {
+        if idx < elementsToSend.count {
+          continuation.resume(returning: elementsToSend[idx])
+        } else {
+          continuation.resume(returning: nil)
+        }
```

Comment:
I don't think we have to do anything with pending elements here. If we have a continuation then we **must not** have any elements buffered. That's a precondition in our internal state. I would be surprised if that can ever happen. I think we just need to loop through all stored continuations and resume them with `nil`

Reply:
ah, thanks for pointing that out – not having to perform this calculation should make things simpler. since i was reasoning locally from the existing implementation, it appeared that case was previously handled, but sounds like that was perhaps an unreachable branch? do you think an actual assertion/precondition of some form is warranted to enforce this invariant (or maybe that's already done and i've overlooked it)?

Reply:
The code here is reasonable complex. Before we assume that this branch is unreachable we should do at least a code read to reason through it. If it truly is unreachable I am happy to put a precondition here since it is only on the finish path.

Reply:
your assessment seems correct to me. the only place where continuations are added is in `next(...)`, and the invariant holds there. the only place pending elements are added is `yield(...)`, and the logic there has 2 cases:

1. if we have stored continuations:
  - by the end of this branch, the total number of continuations will decrease by 1, while the number of stored pending elements will remain unchanged
2. if we have no stored continuations:
  - by the end of this branch, the total number of continuations is unchanged (stays at 0), and we will have added at most 1 new pending element (if the stream isn't terminated)

so i think we can sort of inductively conclude the invariant always holds.

as a consequence, it seems to me the logic in the `yield(...)` method can also likely be simplified, as it appears to contain a dead code path, given this observation.